### PR TITLE
Agrim/DAPI-499/feat: api token table

### PR DESCRIPTION
**Pull request by agrim-deriv**
Implementing the API token table in desktop view, with 6 columns (Token name, Account type, Token, Scopes, Last used, and action items)

### CODE REVIEW

File: src/features/dashboard/components/ApiTokenTable/DeleteTokenDialog/index.tsx [hunk 1]
```diff
@@ -1,57 +1,60 @@
...
+
+  const handleDelete = useCallback(() => {
+    deleteToken(token.token);
+    const newTokens = [];
+    for (let i = 0; i < tokens.length; i++) {
+      if (tokens[i].token !== token.token) {
+        newTokens.push(tokens[i]);
+      }
+    }
+    updateTokens(newTokens);
```
Comment (shafin-deriv) [inline src/features/dashboard/components/ApiTokenTable/DeleteTokenDialog/index.tsx:34]: Do we need to manually handle this from our side? If we make the delete API call shouldn't we get a response with updated token list? If so can we just update the list with that response?

File: src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx [hunk 2]
```diff
@@ -24,110 +25,98 @@ mockUseDeleteToken.mockImplementation(() => ({
...
+  }));
+  render(<ApiTokenTable />);
+});
 
-    const loadingElement = await screen.findByTestId('dt_spinner');
-    expect(loadingElement).toBeVisible();
-  });
+afterEach(() => {
+  jest.clearAllMocks();
 });
```
Comment (shafin-deriv) [inline src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx:56]: Before each and after each should be inside the describe callback

File: src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx [hunk 3]
```diff
@@ -24,110 +25,98 @@ mockUseDeleteToken.mockImplementation(() => ({
...
+  const renderApiTokenTable = () => {
     render(<ApiTokenTable />);
-  });
+  };
 
-  afterEach(() => {
-    cleanup();
-    jest.clearAllMocks();
+  it('Should render all tokens properly', () => {
+    renderApiTokenTable();
```
Comment (shafin-deriv) [inline src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx:64]: what are we exactly expecting here?

File: src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx [hunk 4]
```diff
@@ -24,110 +25,98 @@ mockUseDeleteToken.mockImplementation(() => ({
...
+    const firstActionCell = actionCells[0];
 
-    await userEvent.click(delete_button);
-
-    const delete_modal = await screen.findByText(/Are you sure you want to delete this token?/i);
-    expect(delete_modal).toBeInTheDocument();
+    const withinActionCell = within(firstActionCell);
+    const openDeleteDialogButton = withinActionCell.getByTestId('delete-token-button');
+    await userEvent.click(openDeleteDialogButton);
   });
```
Comment (shafin-deriv) [inline src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx:74]: This test is also without expect!

File: src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx [hunk 5]
```diff
@@ -24,110 +25,98 @@ mockUseDeleteToken.mockImplementation(() => ({
...
-    const close_button = await screen.findByTestId('close-button');
-    await userEvent.click(close_button);
-
-    expect(delete_modal).not.toBeInTheDocument();
+  it('Should close delete dialog on cancel', async () => {
+    const actionCells = await screen.findAllByTestId('token-action-cell');
+    const firstActionCell = actionCells[0];
+    const withinActionCell = within(firstActionCell);
+    const openDeleteDialogButton = withinActionCell.getByTestId('delete-token-button');
+    await userEvent.click(openDeleteDialogButton);
```
Comment (shafin-deriv) [inline src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx:81]: Same here. A test case should always have at least an expect.

File: src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx [hunk 6]
```diff
@@ -24,110 +25,98 @@ mockUseDeleteToken.mockImplementation(() => ({
...
-    const delete_modal = await screen.findByText(/Are you sure you want to delete this token?/i);
-    expect(delete_modal).toBeInTheDocument();
-
-    const cancel_button = await screen.findByRole('button', { name: 'Cancel' });
-    await userEvent.click(cancel_button);
-
-    expect(delete_modal).not.toBeInTheDocument();
+    const withinActionCell = within(firstActionCell);
+    const openDeleteDialogButton = withinActionCell.getByTestId('delete-token-button');
+    await userEvent.click(openDeleteDialogButton);
```
Comment (shafin-deriv) [inline src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx:90]: same as above

File: src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx [hunk 7]
```diff
@@ -24,110 +25,98 @@ mockUseDeleteToken.mockImplementation(() => ({
...
+    const firstActionCell = actionCells[0];
+    const withinActionCell = within(firstActionCell);
 
-    await userEvent.click(delete_button);
-
-    const delete_modal = await screen.findByText(/Are you sure you want to delete this token?/i);
-    expect(delete_modal).toBeInTheDocument();
+    const openDeleteDialogButton = withinActionCell.getByTestId('delete-token-button');
+    await userEvent.click(openDeleteDialogButton);
+  });
```
Comment (shafin-deriv) [inline src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx:100]: same as above

File: src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx [hunk 8]
```diff
@@ -24,110 +25,98 @@ mockUseDeleteToken.mockImplementation(() => ({
...
+    expect(accordion.length).toBe(1);
+  });
 
-    expect(mockDeleteToken).toHaveBeenCalledTimes(1);
-    expect(mockDeleteToken).toHaveBeenCalledWith('first_token');
+  it('Should open first accordion on item click', async () => {
+    renderApiTokenTable();
+    const item = await screen.findAllByText('This is my first token');
+    await userEvent.click(item[0]);
+  });
```
Comment (shafin-deriv) [inline src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx:115]: same as above

File: src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx [hunk 9]
```diff
@@ -24,110 +25,98 @@ mockUseDeleteToken.mockImplementation(() => ({
...
 
-    expect(delete_modal).not.toBeInTheDocument();
+  it('Should update current tab on clicking Create new token button', async () => {
+    renderApiTokenTable();
+    const createTokenButton = screen.getAllByTestId('create-new-token-button');
+    await userEvent.click(createTokenButton[0]);
```
Comment (shafin-deriv) [inline src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx:120]: same as above

File: src/features/dashboard/components/ApiTokenTable/index.tsx [hunk 10]
```diff
@@ -3,19 +3,29 @@ import Spinner from '@site/src/components/Spinner';
...
 import Table from '../Table';
+import { Button, Heading, Text } from '@deriv-com/quill-ui';
+import { LabelPairedCirclePlusMdRegularIcon } from '@deriv/quill-icons';
+import ResponsiveTable from './responsive-table';
+import useDeviceType from '@site/src/hooks/useDeviceType';
 
 export type TTokenColumn = Column<TTokenType>;
 
-const tableColumns: TTokenColumn[] = [
+const tokenTableColumns = (): TTokenColumn[] => [
```
Comment (shafin-deriv) [inline src/features/dashboard/components/ApiTokenTable/index.tsx:20]: I think we don't need to make it a function. The array should be fine to pass directly to the table.

File: src/features/dashboard/components/ApiTokenTable/index.tsx [hunk 11]
```diff
@@ -32,35 +42,60 @@ const tableColumns: TTokenColumn[] = [
...
-  useEffect(() => {
-    if (tokens.length > 0) {
-      setTableHeight(ROW_HEIGHT * tokens.length);
-    }
-  }, [tokens]);
+  const ResponsiveDevice = () => {
+    if (deviceType === 'tablet') return `${styles.api_table} ${styles.tablet}`;
+    if (deviceType === 'mobile') return `${styles.api_table} ${styles.mobile}`;
+    return styles.api_table;
+  };
```
Comment (shafin-deriv) [inline src/features/dashboard/components/ApiTokenTable/index.tsx:69]: We can use `clsx` to achieve the conditional styling. You can look into the `DropdownList` index file for reference.

File: src/features/dashboard/components/ApiTokenTable/responsive-table.scss [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+      font-weight: 500;
+    }
+    &__value {
+      text-align: end;
+      justify-content: end;
+
+      &_row {
+        text-align: start;
+        justify-content: start;
+        overflow-wrap: anywhere;
```
Comment (shafin-deriv) [inline src/features/dashboard/components/ApiTokenTable/responsive-table.scss:35]: is it necessary?

**Comment by shafin-deriv** [comment]
For all the test cases let's ensure we have proper expectations added.

File: src/features/dashboard/components/ApiTokenTable/DeleteTokenDialog/index.tsx [hunk 13]
```diff
@@ -1,57 +1,60 @@
...
-      },
-    ],
-    [setToggleModal],
-  );
+  useEffect(() => {
+    document.body.style.overflow = 'hidden';
+    return () => {
+      document.body.style.overflow = '';
+    };
+  }, []);
```
[the remark this answers is not in the document]
Comment (agrim-deriv) [reply src/features/dashboard/components/ApiTokenTable/DeleteTokenDialog/index.tsx]: Yes, used the useDisableScroll hook

File: src/features/dashboard/components/ApiTokenTable/DeleteTokenDialog/index.tsx (hunk 1 above)
Comment (agrim-deriv) [reply src/features/dashboard/components/ApiTokenTable/DeleteTokenDialog/index.tsx]: Fixed this as well as the updateTokens is part of the response from the API where it's updating the tokens after deletion, this didn't need to be handled manually.

File: src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx (hunk 2 above)
Comment (agrim-deriv) [reply src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx]: Yes, for these tests I've understood the syntax and a bit of the basics. Will keep practicing these to better write these tests.

File: src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx (hunk 3 above)
Comment (agrim-deriv) [reply src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx]: Have changed the tests completely, into somethign that is logical and makes sense. What I did after watching few videos on youtube is looking at tests previously written and the logic of the test.

File: src/features/dashboard/components/ApiTokenTable/index.tsx (hunk 10 above)
Comment (agrim-deriv) [reply src/features/dashboard/components/ApiTokenTable/index.tsx]: Yes @shafin-deriv , correct. Have changed this.

**Review by shafin-deriv — APPROVED** [review]

File: src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx [hunk 14]
```diff
@@ -72,76 +68,31 @@ describe('DeleteTokenDialog', () => {
...
+    const header_descr = await screen.findByText(/Access all your API token details here/i);
+    expect(header_descr).toBeInTheDocument();
   });
 
-  it('Should close the dialog when pressing the close button', async () => {
-    const delete_button = await screen.findByTestId('dt_delete_button_0');
+  it('Shows the dialog when pressing the delete button', async () => {
+    const delete_button = await screen.findByTestId('delete-token-button');
+    await userEvent.click(delete_button);
     expect(delete_button).toBeInTheDocument();
```
Comment (shafin-deriv) [inline src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx:82]: @agrim-deriv Expectation doesn't match with the test purpose here

File: src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx [hunk 15]
```diff
@@ -72,76 +68,31 @@ describe('DeleteTokenDialog', () => {
...
-    const cancel_button = await screen.findByRole('button', { name: 'Cancel' });
-    await act(async () => {
-      await userEvent.click(cancel_button);
-    });
-
-    expect(delete_modal).not.toBeInTheDocument();
+  it('Should delete the token or cancel and close the modal', async () => {
+    const actions_token = await screen.findByTestId('token-action-cell');
+    await userEvent.click(actions_token);
+    expect(actions_token).toBeInTheDocument();
```
Comment (shafin-deriv) [inline src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx:88]: We are performing an action here so we should expect something that should occur after performing the action.

File: src/features/dashboard/components/Table/__tests__/Table.test.tsx [hunk 16]
```diff
@@ -131,11 +136,15 @@ describe('Table', () => {
...
-
+    const actionCells = await screen.findAllByTestId('token-action-cell');
+    const firstActionCell = actionCells[0];
+    const withinActionCell = within(firstActionCell);
+
+<<<<<<< HEAD
+    const openDeleteDialogButton = withinActionCell.getByTestId('delete-token-button');
+    await userEvent.click(openDeleteDialogButton);
+    expect(openDeleteDialogButton).toBeInTheDocument();
+=======
```
Comment (shafin-deriv) [inline src/features/dashboard/components/Table/__tests__/Table.test.tsx:147]: I think the syntax is wrong here.

File: src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx (hunk 15 above)
Comment (agrim-deriv) [reply src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx]: Combined this one and the above one just like how it was done in the apps table test case.

File: src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx (hunk 14 above)
Comment (agrim-deriv) [reply src/features/dashboard/components/ApiTokenTable/__tests__/index.test.tsx]: Combined this test purpose and the below one to combine it similar to how it is used in the apps table test case.

File: src/features/dashboard/components/Table/__tests__/Table.test.tsx (hunk 16 above)
Comment (agrim-deriv) [reply src/features/dashboard/components/Table/__tests__/Table.test.tsx]: Fixed this as well, following how it was done in the apps table test case.

**Review by shafin-deriv — APPROVED** [review]

**Pull request merged**